pkg/metabase/grant.go

```diff
@@ -0,0 +1,397 @@
+	}
+
+	s := strings.Split(subject, ":")
+	if s[0] != "user" {
+		log.Info("subject is not a user")
+		return
+	}
+
+	mbGroupMembers, err := m.client.GetPermissionGroup(ctx, mbMetadata.PermissionGroupID)
+	if err != nil {
+		log.WithError(err).Error("getting permission group")
+		return
+	}
+
+	exists, _ := memberExists(mbGroupMembers, s[1])
+	if exists {
+		log.Info("member already exists")
+		return
+	}
+
+	if err := m.client.AddPermissionGroupMember(ctx, mbMetadata.PermissionGroupID, s[1]); err != nil {
+		log.WithError(err).WithField("user", s).
+			WithField("group", mbMetadata.PermissionGroupID).
+			Warn("Unable to sync user")
+	}
+}
+
+func (m *Metabase) createRestricted(ctx context.Context, ds *models.Dataset) error {
+	groupID, err := m.client.CreatePermissionGroup(ctx, ds.Name)
+	if err != nil {
+		return err
+	}
+
+	aadGroupID, err := m.client.CreatePermissionGroup(ctx, ds.Name+" (aad)")
+	if err != nil {
+		return err
+	}
+
+	// Hack/workaround necessary due to how metabase has implemented saml group sync. When removing a saml group mapping to a
+	// metabase permission group, the users in the saml group will remain members of the permission group.
+	// Adding a dummy mapping ensures that users are evicted from the permission group when an actual group mapping is removed.
+	// See https://github.com/metabase/metabase/issues/26079
+	if err := m.client.UpdateGroupMapping(ctx, "non-existant-aad-group", aadGroupID, GroupMappingOperationAdd); err != nil {
+		return err
+	}
+
+	colID, err := m.client.CreateCollectionWithAccess(ctx, []int{groupID, aadGroupID}, ds.Name)
+	if err != nil {
+		return err
+	}
+
+	key, email, err := m.createServiceAccount(ds)
+	if err != nil {
+		return err
+	}
+
+	err = m.create(ctx, dsWrapper{
+		Dataset:            ds,
+		Key:                string(key),
+		Email:              email,
+		MetabaseGroupID:    groupID,
+		MetabaseAADGroupID: aadGroupID,
+		CollectionID:       colID,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Metabase) create(ctx context.Context, ds dsWrapper) error {
+	datasource, err := m.repo.GetBigqueryDatasource(ctx, ds.Dataset.ID)
+	if err != nil {
+		return err
+	}
+
+	err = m.accessMgr.Grant(ctx, datasource.ProjectID, datasource.Dataset, datasource.Table, "serviceAccount:"+ds.Email)
+	if err != nil {
+		return err
+	}
+
+	dp, err := m.repo.GetDataproduct(ctx, ds.Dataset.DataproductID)
+	if err != nil {
+		return err
+	}
+
+	dbID, err := m.client.CreateDatabase(ctx, dp.Owner.Group, ds.Dataset.Name, ds.Key, ds.Email, &datasource)
+	if err != nil {
+		return err
+	}
+
+	err = m.repo.CreateMetabaseMetadata(ctx, models.MetabaseMetadata{
+		DatasetID:            ds.Dataset.ID,
+		DatabaseID:           dbID,
+		PermissionGroupID:    ds.MetabaseGroupID,
+		AADPermissionGroupID: ds.MetabaseAADGroupID,
+		CollectionID:         ds.CollectionID,
+		SAEmail:              ds.Email,
+	})
+	if err != nil {
+		return err
+	}
+
+	m.waitForDatabase(ctx, dbID, datasource.Table)
```

Comment:
waitForDatabase can get error or timeout, should we handle the situation?

Reply:
We should at least return error and not continue, I've added that now. 

But when the database creation fails we should
- either remove the service account grant, and in the case of restricted databases, clean up the permission groups/collections created and delete the iam service account so that a new metabase mapping from markedsplassen can start from scratch.
- or write the database creation code so that it will reuse service account/permission groups/collections if they already exist.

I think I'm leaning towards the last alternative, do you disagree?

However, I think this could be a separate task not enclosed by this pull request.

Reply:
Agree, and your fix is good enough for this pr